crates/bevy_pbr/src/meshlet/from_mesh.rs

```diff
@@ -19,80 +26,343 @@ impl MeshletMesh {
+        let shared_edge_count = triangle_edges_per_meshlet[&meshlet_id1]
+            .intersection(&triangle_edges_per_meshlet[&meshlet_id2])
+            .count();
+        if shared_edge_count != 0 {
+            connected_meshlets_per_meshlet
+                .get_mut(&meshlet_id1)
+                .unwrap()
+                .push((meshlet_id2, shared_edge_count));
+            connected_meshlets_per_meshlet
+                .get_mut(&meshlet_id2)
+                .unwrap()
+                .push((meshlet_id1, shared_edge_count));
+        }
+    }
+    connected_meshlets_per_meshlet
+}
+
+fn group_meshlets(
+    simplification_queue: Range<usize>,
+    connected_meshlets_per_meshlet: &HashMap<usize, Vec<(usize, usize)>>,
+) -> HashMap<i32, Vec<usize>> {
+    let mut xadj = Vec::with_capacity(simplification_queue.len() + 1);
+    let mut adjncy = Vec::new();
+    let mut adjwgt = Vec::new();
+    for meshlet_id in simplification_queue.clone() {
+        xadj.push(adjncy.len() as i32);
+        for (connected_meshlet_id, shared_edge_count) in
+            connected_meshlets_per_meshlet[&meshlet_id].iter().copied()
+        {
+            adjncy.push((connected_meshlet_id - simplification_queue.start) as i32);
+            adjwgt.push(shared_edge_count as i32);
+        }
+    }
+    xadj.push(adjncy.len() as i32);
+
+    let mut group_per_meshlet = vec![0; simplification_queue.len()];
+    let partition_count = (simplification_queue.len().div_ceil(4)) as i32;
+    Graph::new(1, partition_count, &xadj, &adjncy)
+        .unwrap()
+        .set_adjwgt(&adjwgt)
+        .part_kway(&mut group_per_meshlet)
+        .unwrap();
+
+    let mut groups = HashMap::new();
+    for (i, meshlet_group) in group_per_meshlet.into_iter().enumerate() {
+        groups
+            .entry(meshlet_group)
+            .or_insert(Vec::new())
+            .push(i + simplification_queue.start);
+    }
+    groups
+}
+
+fn simplify_meshlet_groups(
+    group_meshlets: &[usize],
+    meshlets: &Meshlets,
+    vertices: &VertexDataAdapter<'_>,
+    lod_level: u32,
+) -> Option<(Vec<u32>, f32)> {
+    // Build a new index buffer into the mesh vertex data by combining all meshlet data in the group
+    let mut group_indices = Vec::new();
+    for meshlet_id in group_meshlets {
+        let meshlet = meshlets.get(*meshlet_id);
+        for meshlet_index in meshlet.triangles {
+            group_indices.push(meshlet.vertices[*meshlet_index as usize]);
+        }
+    }
+
+    // Allow more deformation for high LOD levels (1% at LOD 1, 10% at LOD 20+)
+    let t = (lod_level - 1) as f32 / 19.0;
+    let target_error = 0.1 * t + 0.01 * (1.0 - t);
+
+    // Simplify the group to ~50% triangle count
+    // TODO: Use simplify_with_locks()
+    let mut error = 0.0;
+    let simplified_group_indices = simplify(
+        &group_indices,
+        vertices,
+        group_indices.len() / 2,
+        target_error,
+        SimplifyOptions::LockBorder,
```

Comment:
So we aren't doing the Nanite thing where they overlap LODs to avoid locking borders, right? We're just locking borders for now?

(I agree with your approach for simplicity's sake, because I never understood what Nanite does. Just wanting to understand)

Reply:
No, we are doing what nanite does. What you want to do is avoid cracks/gaps when displaying different LOD levels for different sections of a mesh.

By grouping meshlets, locking the group border, simplifying the group, and then splitting into new meshlets to form a LOD, you avoid the cracks between LOD levels.

I can't really explain why that works intuitively, but read the nanite slides or the PhD thesis that nanite is based on (link in slides) if you want the theory.